compiler/verification/z3backend.real.ml

```diff
@@ -313,11 +314,17 @@ and find_or_create_enum (ctx : context) (enum : EnumName.t) :
   | Some e -> ctx, e
   | None ->
     let ctrs = EnumMap.find enum ctx.ctx_decl.ctx_enums in
-    let ctx, z3_ctrs = List.fold_left_map create_constructor ctx ctrs in
+    let ctx, z3_ctrs =
+      EnumConstructorMap.fold
+        (fun ctr ty (ctx, ctrs) ->
+          let ctx, ctr = create_constructor ctr ty ctx in
+          ctx, ctr :: ctrs)
+        ctrs (ctx, [])
+    in
     let z3_enum =
       Datatype.mk_sort_s ctx.ctx_z3
         (Marked.unmark (EnumName.get_info enum))
-        z3_ctrs
+        (List.rev z3_ctrs)
```

Comment:
`fold_left` vs `fold_right` semantics for `EnumConstructorMap.fold` ?

Reply:
Yeah this one actually took me a little extra time to debug :/
`Set`/`Map` always fold from the left, and don't even provide reverse fold functions (although they wouldn't be any more expensive...) ; so the fold actually returns a reversed list.

Best solution would probably be to have a `Map.fold_map` function that already preserves order...